### PR TITLE
Refactored Linting & Unit Testing Workflows

### LINKED ISSUE #40285: Consolidate Unit Testing Workflows
These workflows can likely be consolidated into a single workflow:

## Workflows
* `pr-unit-tests.yml`: Runs the PHPUnit test matrix for WooCommerce Core. This is sharded for performance.
* `pr-test-other-plugins.yml`: Runs the unit tests for the WooCommerce Docs plugin.
* `pr-test-js.yml`: Run the unit tests for the WooCommerce Admin client and dependencies.
* `ci.yml`: Runs the PHPUnit tests on `trunk` and `release/**` branches.

By making use of a single Turbo command we are able to run all of the unit tests in a single workflow. The caveat here is that there are a few considerations to keep in mind:

* PHPUnit tests require a `wp-env` environment to run.
* Core's PHPUnit tests are sharded for performance.
* The `ci.yml` workflow runs on the branch after it gets merged.

One thing we can lean into is having a `test:php` and `test:js` script to separate them. This is a good first-step, but, it doesn't solve the need for a `wp-env` environment. One possibility could be leaning into having a single shared `wp-env` environment. PHPUnit tests use a temporary database, so, it's possible that we could use a single database for all of the tests. This would require some investigation. Worst-case we could maybe look at running them concurrently if the GitHub Action UX is acceptable for that. Lastly, for the `ci.yml` workflow, we _could_ probably expand the events to include it, but, if this workflow is something we care about, it might be neat to have a consolidated one that runs linting, unit tests, and E2E tests.

### LINKED ISSUE #40286: Consolidate Linting Workflows
These workflows are really trivially consolidated into a single workflow:

### Workflows
* `pr-lint-js.yml`: Lints all of the JS packages and annotates the code.
* `pr-lint-monorepo.yml`: Checks for changelog files.
* `pr-code-sniff.yml`: Runs PHPCS on WooCommerce Core.
* `pr-lint-markdown.yml`: Lints markdown files in the monorepo.
* `pr-highlight-changes.yml`: Runs the Code Analyzer and highlights changes to template files.
* `changelog-auto-add.yml`: Automatically adds a changelog file based on the PR description template.
* `syncpack.yml`: Checks for dependencies that don't meet the syncpack constraints when a PR is created.

One consideration is that the changelog auto-add can be merged into a single job with the monorepo linting. We only need to check for the changelog file if the auto-add doesn't happen. The rest are pretty easy!

### PR DESCRIPTION
**Pull request by ObliviousHarmony**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This pull request consolidates all of the linting and unit testing workflows in order to make maintenance easier. Having a single workflow allows us to have one central point to optimize as well as provide support to new projects without having to create new workflows. As part of this pull request I am making it so the tasks will only run on projects affected by the pull request. This will improve performance and avoid running unnecessary tasks.

Closes #40285, #40286.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Review the new workflow run in the PR.
2. Review workflows documented in the comments.
3. Create a draft PR from this branch targeting it.
4. Try PHP-only changes
5. Try JS-only changes
6. Try no changes.
7. Try Markdown only changes
8. Try test file only changes (for JS and PHP respectively)
9. Compare generated workflows to existing workflows. This includes WooCommerce Core's testing suite, the other plugin testing suite, and linting workflows. All of the workflows deleted in this PR, basically.

One thing not covered in the testing instructions above is whether or not the workflow works post-merge. I did most of my testing around that locally by looking at the output of the `build-ci-matrix.js` file. When you call `buildCIMatrix()` with a falsey value it adds every project and every type of command that it has.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

Just making some changes to support the new CI workflow.

</details>

### CODE REVIEW

**Comment by ObliviousHarmony** [comment]
Here are two runs of this workflow visible here in this PR:

* JS & PHP Changes: https://github.com/woocommerce/woocommerce/actions/runs/6461108496?pr=40572
  * I introduced a PHPCS linting error here to show that the `phpcs-changed` works.
* WooCommerce Core Change Only: https://github.com/woocommerce/woocommerce/actions/runs/6465680338?pr=40572
  * This necessarily exists because we modified the `package.json` and `composer.json` files for Core.

**Comment by ObliviousHarmony** [comment]
In the future it might make sense to refactor the `testEnv` configuration to support per-command test environments. I can see cases where different types of tests may have different requirements. This seems like it'll probably be necessary when we consolidate the E2E tests. For now though, I think this is ready for review.

**Comment by ObliviousHarmony** [comment]
Test Cases from #40674:

* Documentation Change: https://github.com/woocommerce/woocommerce/actions/runs/6473823011?pr=40674
  * There's currently a workflow that lints markdown but I'm not sure if that linting belongs in this workflow. It makes sense but there is some top-level documentation too. Maybe we need to consider changes outside of a project for the workflow? A problem for tomorrow.
* JS Test File Change: https://github.com/woocommerce/woocommerce/actions/runs/6473886770?pr=40674

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,90 +1,98 @@
...
+          id: 'project-matrix'
+          name: 'Build Matrix'
+          with:
+            script: |
+              let baseRef = ${{ toJson( github.base_ref ) }};
+              if ( baseRef ) {
+                baseRef = 'origin/' + baseRef;
+              }
+              const buildCIMatrix = require( './.github/workflows/scripts/build-ci-matrix' );
+              core.setOutput( 'matrix', JSON.stringify( await buildCIMatrix( baseRef ) ) );
```
Comment (psealock) [inline .github/workflows/ci.yml:40]: Some logging would be useful here.

I introduced [a lint error in `@woocommerce/components`](https://github.com/woocommerce/woocommerce/pull/40691/commits/79422d1c070639564d199b50134ab2904f8b0775) and I do get a failure in JS tests for `woocommerce/client/admin`. I'm trying to debug and see if that's even related. `@woocommerce/components` tests weren't run. See [PR here](https://github.com/woocommerce/woocommerce/pull/40691)

Comment (psealock) [reply .github/workflows/ci.yml:40]: Some logging in the Evaluate Matrix step would be useful too, the error isn't too helpful

Comment (psealock) [reply .github/workflows/ci.yml:40]: Upon restarting, I do get the components to lint, but it passes 🤷🏽 https://github.com/woocommerce/woocommerce/actions/runs/6476659736/job/17588896320?pr=40691

Comment (ObliviousHarmony) [reply .github/workflows/ci.yml:40]: > Some logging would be useful here.

What kind of logging would you be interested in? The output is the matrix, but you can see what jobs were created and the like. The only information that I can think of that might be "hidden" is the intermediary steps leading up to the creation of the matrix. Namely, the files that were changed (and the projects they marked as changed) and the changes that were marked (along with the file that marked them)? This seems like it _could_ be interesting for being able to track how a project got added to the matrix. I'll take a look at adding that.

> Some logging in the Evaluate Matrix step would be useful too, the error isn't too helpful

I'm not really sure what value we could extract from this step. All it does is check for failed jobs so the error indicating that a job has failed is just telling you what you can already see from GitHub's interface on the job list. The `needs` context doesn't provide us information about _which_ jobs failed, but again, you can actually just see that. Once you click on the job you'll see what steps failed.

**Comment by ObliviousHarmony** [comment]
I've pushed another commit that adds logging and more granular detection of asset file changes. This _might_ have created problems with change detection though so that'll need looked at in testing.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (psealock) [reply .github/workflows/ci.yml:40]: >What kind of logging would you be interested in? The output is the matrix, but you can see what jobs were created and the like

I made a change in a particular file which should kick off certain checks, which didn't fire. My attempts to figure out what went wrong (at the matrix creation step? At the matrix interpretation step?) would have been aided by some logging. For instance, a print out of the matrix itself would have let me know that indeed the changes made were correctly picked up (or not) and given me a clue as to where the problem might be. 

I find it useful for tools like this to be overly descriptive. "Here's what I'm dong", "Here's what the result is", and "Here is what I'm going to do next".

Comment (ObliviousHarmony) [reply .github/workflows/ci.yml:40]: I added some pretty extensive logging, take a look and let me know what you think.

**Review by jonathansadowski — APPROVED** [review]
Sorry it took me so long to finish the review of this, I had some false-starts last week 😄 

I've compared the output of the workflow here to the output of the workflows with similar PRs against trunk, and for the parts of the workflow that run, I'm not noticing any difference... and unnecessary tests are not running (no PHP tests for JS changes, and vise-versa).

I've also walked through the changes, as well as the code that generates the matrix, and everything looks good to me, nice work!

I'm approving this, but given its scope, I'd be more comfortable if at least somebody else also approved, but I'll leave that up to you.

**Comment by psealock** [comment]
Thanks for adding the logging, it makes it really transparent whats happening.

For this change in https://github.com/woocommerce/woocommerce/pull/40857, I see `test:js` getting added to the matrix. Perhaps because its detecting a change in `package.json` and `composer.json`? 

image

image

**Comment by ObliviousHarmony** [comment]
Changes to any JSON file are registered as JS source changes @psealock. This catches the MANY cases of JSON config file changes that can impact linting and testing for JS projects. In the future we can get more granular with detection but the heavy-handed approach is at-worst no different than the current one and at-best an improvement when it works

**Comment by psealock** [comment]
JS only change in a component nicely kicks off JS tests in all affected packages and plugins https://github.com/woocommerce/woocommerce/pull/40859

image


image

Interestingly [CI passes](https://github.com/woocommerce/woocommerce/actions/runs/6567854388/job/17841330136?pr=40859), but this passes locally as well - although it shouldn't given the change. This might be something unrelated.

```
pnpm --filter="@woocommerce/components" lint
```

**Comment by psealock** [comment]
>In the future we can get more granular with detection but the heavy-handed approach is at-worst no different than the current one and at-best an improvement when it works

Cool cool, this makes sense. I changed the diff in the same file and got different results, which seem more in line with expectations.

image

image

**Review by psealock — APPROVED** [review]
I've done a fair bit of testing and since any failures anywhere can be gathered post-merge, we should be able to know if anything falls through the cracks.

Nice job with this @ObliviousHarmony !

**Pull request merged**